### PR TITLE
`jax.debug.breakpoint` now supports being DCE'd. Drive-by: fix #16186

### LINKED ISSUE #16186: `jax.debug.breakpoint` crashes with numpy arrays of objects.
### Description

```python
import jax
import numpy as np

@jax.jit
def f():
    y = np.array([object(), object()])
    jax.debug.breakpoint()

f()
# AssertionError: [<object object at 0x7d3e8ddab2c0> <object object at 0x7d3e8cbc8510>]
```
It looks like `jax.debug.breakpoint` is grabbing all the arraylikes from the stack, but is falsely assuming that all NumPy arrays can be treated as a jaxtype.

### What jax/jaxlib version are you using?

HEAD

### PR DESCRIPTION
**Pull request by patrick-kidger**
As discussed internally.

Tagging @sharadmv for review.

### CODE REVIEW

File: jax/_src/debugger/core.py [hunk 1]
```diff
@@ -84,7 +84,9 @@ def tree_flatten(self):
     flat_globals, globals_tree = _safe_flatten_dict(self.globals)
     flat_vars = flat_locals + flat_globals
     is_valid = [
-        isinstance(l, (core.Tracer, jax.Array, np.ndarray))
+        isinstance(l, (core.Tracer, jax.Array, np.ndarray)) and
+        (jax.numpy.issubdtype(l.dtype, jax.numpy.bool_) or
+         jax.numpy.issubdtype(l.dtype, jax.numpy.number))
```
Comment (patrick-kidger) [inline jax/_src/debugger/core.py:89]: Actually, I think these are the tracers passed to the primitive bind. Should this just be `isinstance(l, core.Tracer)`? Everything else is valid as a static argument.

File: jax/_src/debugger/core.py [hunk 2]
```diff
@@ -207,4 +213,12 @@ def _breakpoint_callback(*flat_args):
...
-  debugging.debug_callback(_breakpoint_callback, *flat_args, ordered=ordered)
+  if token is None:
+    debugging.debug_callback(_breakpoint_callback, *flat_args, ordered=ordered)
+  else:
+    def _breakpoint_callback_wrapper(x, *flat_args):
+      _breakpoint_callback(*flat_args)
+      return x
+    token, flat_args = jax.lax.stop_gradient((token, flat_args))
+    return jax.pure_callback(_breakpoint_callback_wrapper, token, token, *flat_args,
+                             vectorized=True)
```
Comment (sharadmv) [inline jax/_src/debugger/core.py:224]: I don't think this matches semantics with the debug callback version. Any reason you're setting vectorized=True here?

Comment (patrick-kidger) [reply jax/_src/debugger/core.py:224]: Agreed this doesn't match the semnatics. I can revert that change for now.

But longer-term I'd like to either add a `vectorized` flag to `jax.debug.breakpoint`, or unconditionally make the `debug_callback` path be vectorized as well. I find it to be far more useful, especally when working with large batches.

**Review by sharadmv — APPROVED** [review]
Thanks!

**Pull request merged**